Add configuration option to send full test name paths to junit 5 platform

The [JunitXmlReporter](https://kotest.io/docs/extensions/junit_xml.html) extension enables getting longer test names in an xml file, but surefire's console output still uses the shorter names.  It would be great to have the longer names in the surefire output as well.

Ok, let me know if this works.  The output of the failed test looks like this:
```
[INFO]
[ERROR] Failures:
[ERROR] org.jitsi.nlj.util.Rfc3711IndexTrackerTest.should return the proper index
[INFO]   Run 1: PASS
[INFO]   Run 2: PASS
[ERROR]   Run 3: Rfc3711IndexTrackerTest expected:<64001> but was:<64000>
```

But the test itself is:
```
  internal class Rfc3711IndexTrackerTest : ShouldSpec() {
      override fun isolationMode(): IsolationMode? = IsolationMode.InstancePerLeaf

      private val indexTracker = Rfc3711IndexTracker()

      init {
          context("feeding in the first sequence number") {
              val firstIndex = indexTracker.update(65000)
              should("return itself as the index") {
                  firstIndex shouldBe 65000
              }
              context("and then another without rolling over") {
                  val secondIndex = indexTracker.update(65001)
                  should("return itself as the index") {
                      secondIndex shouldBe 65001
                  }
                  context("and then another which does roll over") {
                      val rollOverIndex = indexTracker.update(2)
                      should("return the proper index") {
                          rollOverIndex shouldBe 1 /* roc */ * 0x1_0000 + 2L
                      }
                      context("and then a sequence number from the previous rollover") {
                          val prevRollOverIndex = indexTracker.update(65002)
                          should("return the proper index") {
                              prevRollOverIndex shouldBe 65002
                          }
                      }
                  }
                  context("and then an older sequence number") {
                      val oldIndex = indexTracker.update(64000)
                      should("return the proper index") {
                          // failing test is here
                          oldIndex shouldBe oldIndex + 1
                      }
                  }
              }
          }
          context("a series of sequence numbers") {
              should("never return a negative index") {
                  var seqNum = 22134
                  repeat(35537) {
                      indexTracker.update(seqNum++) shouldBeGreaterThan 0
                  }
              }
          }
      }
  }
```

So the idea was that, in the failure output, it would show the entire test description "chain" (i.e. `Rfc3711IndexTrackerTest feeding in the first sequence number and then another without rolling over and then an older sequence number should return the proper index`)

--- a/kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/config/Configuration.kt
+++ b/kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/config/Configuration.kt
@@ -317,6 +317,8 @@ class Configuration {
     */
    var duplicateTestNameMode: DuplicateTestNameMode = Defaults.duplicateTestNameMode
 
+   var displayFullTestPath: Boolean = Defaults.displayFullTestPath
+
    /**
     * Returns all globally registered [Listener]s.
     */

--- a/kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/config/Defaults.kt
+++ b/kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/config/Defaults.kt
@@ -15,6 +15,7 @@ object Defaults {
    val testCaseOrder: TestCaseOrder = TestCaseOrder.Sequential
    val isolationMode: IsolationMode = IsolationMode.SingleInstance
    val duplicateTestNameMode: DuplicateTestNameMode = DuplicateTestNameMode.Warn
+   val displayFullTestPath: Boolean = false
 
    const val coroutineDebugProbes: Boolean = false
    const val displaySpecIfNoActiveTests: Boolean = true

--- a/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/names/DefaultDisplayNameFormatter.kt
+++ b/kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/names/DefaultDisplayNameFormatter.kt
@@ -23,7 +23,9 @@ fun getDisplayNameFormatter(configuration: Configuration): DisplayNameFormatter 
  * A default implementation of [DisplayNameFormatter].
  * Used when there are no registered [io.kotest.core.extensions.DisplayNameFormatterExtension]s.
  */
-class DefaultDisplayNameFormatter(private val configuration: Configuration) : DisplayNameFormatter {
+class DefaultDisplayNameFormatter(
+   private val configuration: Configuration,
+) : DisplayNameFormatter {
 
    override fun format(testCase: TestCase): String {
 
@@ -48,10 +50,15 @@ class DefaultDisplayNameFormatter(private val configuration: Configuration) : Di
          }
       }
 
-      return if (configuration.testNameAppendTags) {
+      val name = if (configuration.testNameAppendTags) {
          return appendTagsInDisplayName(testCase, displayName)
       } else {
          displayName
+      }
+
+      return when (val parent = testCase.parent) {
+         null -> name
+         else -> if (configuration.displayFullTestPath) format(parent) + " " + name else name
       }
    }
 

--- a/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/names/DefaultDisplayNameFormatterTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/names/DefaultDisplayNameFormatterTest.kt
@@ -21,7 +21,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
    init {
 
       test("@DisplayName should be used for spec name") {
-         DefaultDisplayNameFormatter(configuration).format(SpecWithDisplayName::class) shouldBe "ZZZZZ"
+         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(SpecWithDisplayName::class) shouldBe "ZZZZZ"
       }
 
       test("tags should be appended from spec when configuration is set") {
@@ -35,7 +35,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             sourceRef(),
             TestType.Test,
          )
-         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Foo]"
+         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Foo]"
          configuration.testNameAppendTags = false
       }
 
@@ -51,7 +51,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             type = TestType.Test,
             config = TestCaseConfig(tags = setOf(Dummy, NoUse))
          )
-         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Dummy, NoUse]"
+         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Dummy, NoUse]"
          configuration.testNameAppendTags = false
       }
 
@@ -67,7 +67,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             type = TestType.Test,
             config = TestCaseConfig(tags = setOf(Dummy, NoUse))
          )
-         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Dummy, NoUse, Foo]"
+         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Dummy, NoUse, Foo]"
          configuration.testNameAppendTags = false
       }
    }

--- a/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/names/DefaultDisplayNameFormatterTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/com/sksamuel/kotest/engine/test/names/DefaultDisplayNameFormatterTest.kt
@@ -2,6 +2,7 @@ package com.sksamuel.kotest.engine.test.names
 
 import io.kotest.core.Tag
 import io.kotest.core.annotation.Tags
+import io.kotest.core.config.Configuration
 import io.kotest.core.config.configuration
 import io.kotest.core.descriptors.append
 import io.kotest.core.descriptors.toDescriptor
@@ -21,7 +22,30 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
    init {
 
       test("@DisplayName should be used for spec name") {
-         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(SpecWithDisplayName::class) shouldBe "ZZZZZ"
+         DefaultDisplayNameFormatter(configuration).format(SpecWithDisplayName::class) shouldBe "ZZZZZ"
+      }
+
+      test("test name should use full path option") {
+         val conf = Configuration()
+         conf.displayFullTestPath = true
+         val tc1 = TestCase(
+            SpecWithTag::class.toDescriptor().append("test"),
+            TestName("test"),
+            SpecWithTag(),
+            {},
+            sourceRef(),
+            TestType.Test,
+         )
+         val tc2 = TestCase(
+            SpecWithTag::class.toDescriptor().append("test2"),
+            TestName("test2"),
+            SpecWithTag(),
+            {},
+            sourceRef(),
+            TestType.Test,
+            parent = tc1
+         )
+         DefaultDisplayNameFormatter(conf).format(tc2) shouldBe "test test2"
       }
 
       test("tags should be appended from spec when configuration is set") {
@@ -35,7 +59,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             sourceRef(),
             TestType.Test,
          )
-         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Foo]"
+         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Foo]"
          configuration.testNameAppendTags = false
       }
 
@@ -51,7 +75,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             type = TestType.Test,
             config = TestCaseConfig(tags = setOf(Dummy, NoUse))
          )
-         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Dummy, NoUse]"
+         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Dummy, NoUse]"
          configuration.testNameAppendTags = false
       }
 
@@ -67,7 +91,7 @@ class DefaultDisplayNameFormatterTest : FunSpec() {
             type = TestType.Test,
             config = TestCaseConfig(tags = setOf(Dummy, NoUse))
          )
-         DefaultDisplayNameFormatter(configuration, configuration.displayFullTestPath).format(tc) shouldBe "test[tags = Dummy, NoUse, Foo]"
+         DefaultDisplayNameFormatter(configuration).format(tc) shouldBe "test[tags = Dummy, NoUse, Foo]"
          configuration.testNameAppendTags = false
       }
    }

--- a/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
+++ b/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/JUnitTestEngineListener.kt
@@ -1,6 +1,6 @@
 package io.kotest.runner.junit.platform
 
-import io.kotest.core.config.configuration
+import io.kotest.core.config.Configuration
 import io.kotest.core.descriptors.Descriptor
 import io.kotest.core.descriptors.DescriptorId
 import io.kotest.core.descriptors.spec
@@ -72,6 +72,7 @@ import kotlin.reflect.KClass
 class JUnitTestEngineListener(
    private val listener: EngineExecutionListener,
    val root: EngineDescriptor,
+   private val configuration: Configuration,
 ) : TestEngineListener {
 
    companion object {

--- a/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestJunitPlatformTestEngine.kt
+++ b/kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/KotestJunitPlatformTestEngine.kt
@@ -60,7 +60,9 @@ class KotestJunitPlatformTestEngine : TestEngine {
             JUnitTestEngineListener(
                SynchronizedEngineExecutionListener(
                   request.engineExecutionListener
-               ), root
+               ),
+               root,
+               configuration,
             )
          )
       )

--- a/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/JUnitTestEngineListenerTest.kt
+++ b/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/JUnitTestEngineListenerTest.kt
@@ -1,5 +1,7 @@
 package com.sksamuel.kotest.runner.junit5
 
+import io.kotest.core.config.Configuration
+import io.kotest.core.config.configuration
 import io.kotest.core.descriptors.append
 import io.kotest.core.descriptors.toDescriptor
 import io.kotest.core.names.TestName
@@ -52,7 +54,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an error before spec started should show spec with a dummy error test") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specExit(MySpec::class, Exception("CRRAACK"))
       track.events shouldBe listOf(
@@ -79,7 +81,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an ignored spec should generate no events") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specIgnored(MySpec::class)
       listener.specExit(MySpec::class, null)
@@ -88,7 +90,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an inactive spec should be marked as skipped") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specInactive(
          MySpec::class,
@@ -103,7 +105,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("a successful root test should be marked as started and finished") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -121,7 +123,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("a failed root test should be marked as FAILED") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -140,7 +142,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an ignored root test should be marked as skipped") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testIgnored(tc1, "secret!")
@@ -156,7 +158,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("a successful nested test should be marked as SUCCESSFUL with type TEST") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -183,7 +185,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("a failed nested test should be marked as FAILED with type TEST") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -210,7 +212,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an ignored nested test should be marked as skipped") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -235,7 +237,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("an error in the spec should add a placeholder test with the error along with completed tests") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -266,7 +268,7 @@ class JUnitTestEngineListenerTest : FunSpec({
 
    test("state should be reset after spec") {
       val track = EventTrackingEngineExecutionListener()
-      val listener = JUnitTestEngineListener(track, root)
+      val listener = JUnitTestEngineListener(track, root, configuration)
       listener.specEnter(MySpec::class)
       listener.specStarted(MySpec::class)
       listener.testStarted(tc1)
@@ -306,6 +308,35 @@ class JUnitTestEngineListenerTest : FunSpec({
             "com.sksamuel.kotest.runner.junit5.MySpec2",
             TestExecutionResult.Status.SUCCESSFUL
          ),
+      )
+   }
+
+   test("listener should support full test paths") {
+      val track = EventTrackingEngineExecutionListener()
+      val config = Configuration()
+      config.displayFullTestPath = true
+      val listener = JUnitTestEngineListener(track, root, config)
+      listener.specEnter(MySpec::class)
+      listener.specStarted(MySpec::class)
+      listener.testStarted(tc1)
+      listener.testStarted(tc2)
+      listener.testFinished(tc2, TestResult.success(3))
+      listener.testFinished(tc1, TestResult.success(7))
+      listener.specFinished(
+         MySpec::class,
+         mapOf(tc1 to TestResult.success(7), tc2 to TestResult.ignored("secret!"))
+      )
+      listener.specExit(MySpec::class, null)
+      track.events shouldBe listOf(
+         EventTrackingEngineExecutionListener.Event.TestRegistered("com.sksamuel.kotest.runner.junit5.MySpec", TestDescriptor.Type.CONTAINER),
+         EventTrackingEngineExecutionListener.Event.ExecutionStarted("com.sksamuel.kotest.runner.junit5.MySpec"),
+         EventTrackingEngineExecutionListener.Event.TestRegistered("foo", TestDescriptor.Type.CONTAINER),
+         EventTrackingEngineExecutionListener.Event.ExecutionStarted("foo"),
+         EventTrackingEngineExecutionListener.Event.TestRegistered("foo bar", TestDescriptor.Type.TEST),
+         EventTrackingEngineExecutionListener.Event.ExecutionStarted("foo bar"),
+         EventTrackingEngineExecutionListener.Event.ExecutionFinished("foo bar", TestExecutionResult.Status.SUCCESSFUL),
+         EventTrackingEngineExecutionListener.Event.ExecutionFinished("foo", TestExecutionResult.Status.SUCCESSFUL),
+         EventTrackingEngineExecutionListener.Event.ExecutionFinished("com.sksamuel.kotest.runner.junit5.MySpec", TestExecutionResult.Status.SUCCESSFUL),
       )
    }
 })

--- a/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/JUnitTestRunnerListenerTest.kt
+++ b/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/JUnitTestRunnerListenerTest.kt
@@ -1,5 +1,6 @@
 package com.sksamuel.kotest.runner.junit5
 
+import io.kotest.core.config.configuration
 import io.kotest.core.descriptors.append
 import io.kotest.core.sourceRef
 import io.kotest.core.spec.style.DescribeSpec
@@ -64,7 +65,7 @@ class JUnitTestRunnerListenerTest : DescribeSpec({
             parent = test1,
          )
 
-         val listener = JUnitTestEngineListener(engineListener, root)
+         val listener = JUnitTestEngineListener(engineListener, root, configuration)
          listener.engineStarted(emptyList())
          listener.specEnter(JUnitTestRunnerListenerTest::class)
          listener.specStarted(JUnitTestRunnerListenerTest::class)

--- a/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/SpecInitializationErrorTest.kt
+++ b/kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/com/sksamuel/kotest/runner/junit5/SpecInitializationErrorTest.kt
@@ -1,5 +1,6 @@
 package com.sksamuel.kotest.runner.junit5
 
+import io.kotest.core.config.configuration
 import io.kotest.core.spec.style.FunSpec
 import io.kotest.engine.concurrency.NoopCoroutineDispatcherFactory
 import io.kotest.engine.spec.ReflectiveSpecRef
@@ -37,7 +38,7 @@ class SpecInitializationErrorTest : FunSpec({
          override fun dynamicTestRegistered(testDescriptor: TestDescriptor?) {}
       }
 
-      val listener = JUnitTestEngineListener(engineListener, root)
+      val listener = JUnitTestEngineListener(engineListener, root, configuration)
       val executor = SpecExecutor(listener, NoopCoroutineDispatcherFactory)
       executor.execute(ReflectiveSpecRef(SpecWithFieldError::class))
 
@@ -69,7 +70,7 @@ class SpecInitializationErrorTest : FunSpec({
          override fun dynamicTestRegistered(testDescriptor: TestDescriptor?) {}
       }
 
-      val listener = JUnitTestEngineListener(engineListener, root)
+      val listener = JUnitTestEngineListener(engineListener, root, configuration)
       val executor = SpecExecutor(listener, NoopCoroutineDispatcherFactory)
       executor.execute(ReflectiveSpecRef(SpecWithInitError::class))
 

